### PR TITLE
[ticket/15879] Add core.posting_modify_uninit_data

### PR DESCRIPTION
**Pull request by toxyy**
Allows the modification of default variables put into post_data.
Also allows you to unset variables in post_data to inherit the default ones.

PHPBB3-15879

Checklist:

- [x] Correct branch: master for new features; 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.2.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-15879

### CODE REVIEW

**Comment by toxyy** [comment]
!set Event

File: phpBB/posting.php [hunk 1]
```diff
@@ -606,9 +606,24 @@
...
+*/
+$vars = array(
+	'post_data',
+	'attachment_user_id',
+);
+extract($phpbb_dispatcher->trigger_event('core.posting_modify_attachment_user_id', compact($vars)));
+
 // Always check if the submitted attachment data is valid and belongs to the user.
 // Further down (especially in submit_post()) we do not check this again.
-$message_parser->get_submitted_attachment_data($post_data['poster_id']);
```
Comment (3D-I) [inline phpBB/posting.php]: I'd move the event underneath this line and add the `message_parser` to the array.

Then undo the changes to the `message_parser` 's line of code.

**Comment by 3D-I** [comment]
Wrong header for the commit `add $post_data to event ` - should be like the first one.

File: phpBB/posting.php (hunk 1 above)
Comment (toxyy) [reply phpBB/posting.php]: Unfortunately if `get_submitted_attachment_data` is called at all with an unmodified poster_id in my situation, users would still get an error.  I attempted to move it and add the message parser and re execute it under my conditions, it doesn't bypass the error.

**Comment by toxyy** [comment]
I'll change it once the review you made is resolved.

File: phpBB/posting.php (hunk 1 above)
Comment (3D-I) [reply phpBB/posting.php]: Without testing I'd go that way, more or less.

```
$message_parser->get_submitted_attachment_data($post_data['poster_id']);

/**
* This event allows you to modify the post data etc..
*
* @event core.posting_modify_post_data
* @var	array	post_data		Array with post data
* @var	@todo	message_parser	The message parser etc..
* @since 3.2.5-RC1
*/
$vars = array(
	'post_data',
	'message_parser',
);
extract($phpbb_dispatcher->trigger_event('core.posting_modify_post_data', compact($vars)));

```

and in the event
```

$message_parser = event['message_parser'];
$poster_id = event['post_data']['poster_id'];

$my_id = do something to define your user ID;
$poster_id = (int) $my_id;

$message_parser->get_submitted_attachment_data($poster_id);

event['message_parser'] = event['message_parser'];

```

Just to throw something on the table though.

Comment (toxyy) [reply phpBB/posting.php]: That's pretty much exactly what I did after your initial suggestion.  Since `$message_parser->get_submitted_attachment_data($post_data['poster_id'])` is ran with the wrong poster_id initially, it throws the error and doesn't give the message_parser a second chance to run.

Comment (toxyy) [reply phpBB/posting.php]: image

See the error, it gets automatically called at line 614 (the first call to get_submitted_attachment_data), and nothing in the event can be executed.

Comment (3D-I) [reply phpBB/posting.php]: Reading the code a bit more, the `poster_id` is being set here (` $user->data['user_id']` )

https://github.com/phpbb/phpbb/blob/838b5159acf3a10168b28c23226ff229b7960c90/phpBB/posting.php#L597

Why don't you try to use that array  `uninit` with a new event?

Comment (toxyy) [reply phpBB/posting.php]: I'll test doing that tomorrow, as I'm unsure of the affect of letting the remainder of posting.php process the post with a modified poster_id.  I'd like to keep it ANONYMOUS as long as I can, until I might change it in other events.  Nevertheless I'll attempt this tomorrow and see what happens.

Comment (toxyy) [reply phpBB/posting.php]: Seems to be working just fine, and I suppose this event is more useful than the one I made anyways.  Thanks for your suggestions and help.

**Comment by 3D-I** [comment]
The problem here will be you should rebase (and possibly cherry-pick) the PR/commits since there is always the wrong header for the prior commit `add $post_data to event` 

If you are not confortable with that I suggest you to close this PR and open a new one.
It is faster, nonetheless.

**Comment by 3D-I** [comment]
In fact
```

Error: heading should be a sentence beginning with a letter. 
You entered: 
add $post_data to event 
Syntax error on line 1: 
>> add $post_data to event 
Expecting: comment, header, 
Aborting commit. 
```
